spec/features/webauthn/sign_in_spec.rb

```diff
@@ -49,4 +49,19 @@
     expect(page).to have_content(t('errors.general'))
     expect(page).to have_current_path(login_two_factor_webauthn_path)
   end
+
+  it 'does not show error after successful challlenge/secret reattempt' do
+    mock_webauthn_verification_challenge
+
+    sign_in_user(webauthn_configuration.user)
+    # click the next button or cancel from the browser dialog
+    click_button t('forms.buttons.continue')
+
+    expect(page).to have_content(t('errors.general'))
+
+    mock_press_button_on_hardware_key_on_verification
+    click_button t('forms.buttons.continue')
+
+    expect(page).to_not have_content(t('errors.general'))
```

Comment:
I spent part of my afternoon yesterday exploring if we could have more realistic WebAuthn stubbing for JavaScript-enabled browsers in the spec. I think it's possible, but I didn't quite reach a working solution.

Two avenues I considered:

- Replacing the global [`navigator.credentials.get`](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/get) and [`navigator.credentials.create`](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/create) with a simulated result
- Using the Chrome DevTools Protocol to issue [WebAuthn "bridge" commands](https://chromedevtools.github.io/devtools-protocol/tot/WebAuthn/) (see [BrowserEmulationHelper](https://github.com/18F/identity-idp/blob/main/spec/support/features/browser_emulation_helper.rb) as an example of where we've used DevTools Protocol in the past)

I'd imagine the logic for these could live in the underlying helpers, e.g. [`mock_press_button_on_hardware_key_on_verification`](https://github.com/18F/identity-idp/blob/39999988e0099d67d1a1661a1840916065f642ac/spec/support/features/webauthn_helper.rb#L48-L57), using different code paths depending if `javascript_enabled?` is `true`.

I like the WebAuthn DevTools approach, but I couldn't quite figure out the right parameters to send, and which methods to call. The closest I got to a working solution was implementing the first approach. I got the tests passing locally, but strangely only when the browser was shown (i.e. using the `SHOW_BROWSER=true` environment variable). It's also kinda unwieldy because it inlines parts of [`base64ToArrayBuffer`](https://github.com/18F/identity-idp/blob/39999988e0099d67d1a1661a1840916065f642ac/app/javascript/packages/webauthn/converters.ts#L1-L13).

```rb
def mock_press_button_on_hardware_key_on_verification
  # this is required because the domain is embedded in the supplied attestation object
  allow(WebauthnSetupForm).to receive(:domain_name).and_return('localhost:3000')

  if javascript_enabled?
    credential = {
      rawId: credential_id,
      response: {
        authenticatorData: authenticator_data,
        clientDataJSON: verification_client_data_json,
        signature: signature,
      },
    }
    page.evaluate_script(<<~JS)
      mapObjectValuesToArrayBuffer = (obj) => Object.fromEntries(
        Object.entries(obj)
          .map(([key, value]) => [
            key,
            typeof value === 'string' ?
              Uint8Array.from(
                window
                  .atob(value)
                  .split('')
                  .map((char) => char.charCodeAt(0)),
              ).buffer :
              mapObjectValuesToArrayBuffer(value),
          ])
      )
    JS
    page.evaluate_script(<<~JS)
      navigator.credentials.get = () => Promise.resolve(
        mapObjectValuesToArrayBuffer(#{credential.to_json}),
      )
    JS
    click_button t('two_factor_authentication.webauthn_platform_use_key')
    expect(page).to have_button(t('forms.buttons.continue'), wait: 5)
  else
    # simulate javascript that is triggered when the hardware key button is pressed
    set_hidden_field('credential_id', credential_id)
    set_hidden_field('authenticator_data', authenticator_data)
    set_hidden_field('signature', signature)
    set_hidden_field('client_data_json', verification_client_data_json)
  end
end
```

Long story short: I worry this test isn't really giving us any regression assurances, since I suspect it passes on `main` as well. We could alternatively try a JavaScript spec which tests the JavaScript code. That would also be rather difficult with how `packs` are tightly integrated with page markup and very side-effecty. Maybe in the future we'd refactor the authentication button to a custom element like `app/javascript/packages/webauthn/webauthn-authenticate-button-element.ts`, which could be easier to test against.

Given all the challenges, I think it'd be okay to leave it like this for now, though maybe we should create a ticket to further explore one of the above alternatives?

Reply:
I will work on creating that ticket.

I can confirm that test does pass on main so your suspicion is correct.